File: solutions/php/robot-name/1/RobotName.php
<?php

declare(strict_types=1);

class Robot
{
    static $issuedNames = [];
    private $name = null;

    static function getRandomLetter(): string
    {
        return chr(rand(ord('A'), ord('Z')));
    }

    public function getName(): string
    {
        if ($this->name == null) {
            do {
                $candidate = sprintf("%s%s%'.03d", Robot::getRandomLetter(), Robot::getRandomLetter(), rand(0, 999));
            } while (in_array($candidate, Robot::$issuedNames));

            $this->name = $candidate;
            array_push(Robot::$issuedNames, $candidate);
        }

        return $this->name;
    }

    public function reset(): void
    {
        $this->name = null;
    }
}

File: solutions/python/robot-name/1/robot_name.py
import random


class Robot:
    used_names = set()
    
    @staticmethod
    def random_letter():
        return chr(random.randint(ord('A'), ord('Z')))
    
    @staticmethod
    def random_number():
        return random.randint(0, 9)

    @classmethod
    def generate_name(cls):
        while True:
            candidate = f"{cls.random_letter()}{cls.random_letter()}{cls.random_number()}{cls.random_number()}{cls.random_number()}"
            if candidate not in cls.used_names:
                cls.used_names.add(candidate)
                return candidate
        
    def __init__(self):
        self.name = self.generate_name()

    def reset(self):
        self.name = self.generate_name()
        